### PR TITLE
[Nested Types] Test coverage, pyarrow 7 support, misc. fixes

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds more thorough test coverage for nested types, adds pyarrow 7 support, and includes misc. fixes required to make the added tests pass.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -5,7 +5,7 @@ prettytable-rs = "^0.10"
 rand = "^0.8"
 
 [dependencies.arrow2]
-features = ["compute"]
+features = ["compute", "io_ipc"]
```
Comment (clarkzinzow) [inline Cargo.toml:8]: Required for the IPC roundtrip workaround for the struct/fixed-size list array bug(s) on the arrow2 --> pyarrow FFI boundary.

File: src/ffi.rs [hunk 2]
```diff
@@ -118,3 +131,60 @@ pub fn table_to_record_batch(table: &Table, py: Python, pyarrow: &PyModule) -> P
...
+    -- The Workaround --
+
+    This helper ensures that such offsets are eliminated by ensuring that the underlying data
+    buffers are truncated to the slice; note that this creates a copy of the underlying data,
+    so FFI is not currently zero-copy for struct arrays or fixed-size list arrays. We accomplish
+    this buffer truncation by doing an IPC roundtrip on the array, which should result in a single
+    copy of the array's underlying data.
+    */
+    // Write the IPC representation to an in-memory buffer.
+    // TODO(Clark): Preallocate the vector with the requisite capacity, based on the array size?
```
Comment (clarkzinzow) [inline src/ffi.rs:160]: Could do something like `(array.values_iter().map(|arr| mem::size_of_val(arr)).sum() + mem::size_of_val(array.valdity()) + schema_buffer) / 8`, maybe?

File: src/ffi.rs [hunk 3]
```diff
@@ -118,3 +131,60 @@ pub fn table_to_record_batch(table: &Table, py: Python, pyarrow: &PyModule) -> P
...
+    let stream_metadata = arrow2::io::ipc::read::read_stream_metadata(&mut cursor).unwrap();
+    let mut reader = arrow2::io::ipc::read::StreamReader::new(cursor, stream_metadata, None);
+    // There should only be a single chunk in the stream.
+    let state = reader.next().unwrap();
+    assert!(reader.next().is_none());
+    // Stream should be finished from the reader's perspective.
+    assert!(reader.is_finished());
+    match state {
+        Ok(arrow2::io::ipc::read::StreamState::Some(chunk)) => chunk.arrays()[0].clone(),
+        _ => panic!("shouldn't be reached"),
```
Comment (jaychia) [inline src/ffi.rs:188]: Is some variant of `state.unwrap().unwrap()` possible? 😛

File: src/ffi.rs [hunk 4]
```diff
@@ -87,10 +89,21 @@ pub fn to_py_array(array: ArrayRef, py: Python, pyarrow: &PyModule) -> PyResult<
         array.data_type().clone(),
         true,
     )));
-    let array = Box::new(ffi::export_array_to_c(array));
+    let mut new_arr = array;
+    // TODO(Clark): Fix struct array and fixed-size list array slice FFI upstream in arrow2/pyarrow.
```
Comment (jaychia) [inline src/ffi.rs:93]: Should we move this logic into `fix_child_array_slice_offsets`? I'd imagine we may need more "fixer" functions in the future

**Review by jaychia — APPROVED** [review]
Great tests!

File: src/ffi.rs (hunk 3 above)
Comment (clarkzinzow) [reply src/ffi.rs:188]: I briefly tried, but the intermediate `StreamState::Some` enum variant makes it difficult, but I can try to find a linear `unwrap()` chain real quick.

**Pull request merged**